src/stpipe/library.py

```diff
@@ -0,0 +1,850 @@
+
+    For ease-of-use this ledger maintains 2 mappings:
+
+        - id (the id(model) result) to model index
+        - index to model
+
+    The "index to model" mapping keeps a reference to every
+    model in the ledger (which allows id(model) to be consistent).
+
+    The ledger is a MutableMapping that supports look up of:
+        - index for a model
+        - model for an index
+    """
+
+    def __init__(self):
+        self._id_to_index = {}
+        self._index_to_model = {}
+
+    def __getitem__(self, model_or_index):
+        if isinstance(model_or_index, int):
+            return self._index_to_model[model_or_index]
+        return self._id_to_index[id(model_or_index)]
+
+    def __setitem__(self, index, model):
+        self._index_to_model[index] = model
+        self._id_to_index[id(model)] = index
+
+    def __delitem__(self, model_or_index):
+        if isinstance(model_or_index, int):
+            index = model_or_index
+            model = self._index_to_model[index]
+        else:
+            model = model_or_index
+            index = self._id_to_index[id(model)]
+        del self._id_to_index[id(model)]
+        del self._index_to_model[index]
+
+    def __iter__(self):
+        # only return indexes
+        return iter(self._index_to_model)
+
+    def __len__(self):
+        return len(self._id_to_index)
+
+
+class AbstractModelLibrary(abc.ABC):
+    """
+    A "library" of models (loaded from an association file).
+
+    Do not anger the librarian!
+
+    The library owns all models from the association and it will handle
+    opening and closing files.
+
+    Models can be "borrowed" from the library (by iterating through the
+    library or "borrowing" a specific model). However the library must be
+    "open" (used in a ``with`` context)  to borrow a model and the model
+    must be "shelved" before the library "closes" (the ``with`` context exits).
+
+    >>> with library:   # doctest: +SKIP
+            model = library.borrow(0)  # borrow the first model
+            # do stuff with the model
+            library.shelve(model, 0)  # return the model
+
+    Failing to "open" the library will result in a `ClosedLibraryError`.
+
+    Failing to "return" a borrowed model will result in a `BorrowError`.
+    """
+
+    def __init__(
+        self,
+        init,
+        asn_exptypes=None,
+        asn_n_members=None,
+        on_disk=False,
+        temp_directory=None,
+        **datamodels_open_kwargs,
+    ):
+        """
+        Create a new ModelLibrary based on the provided "init".
+
+        Parameters
+        ----------
+        init : str or Path or list (of DataModels)
+            If a string or Path this should point to a valid
+            association file. If a list of models consider continuing
+            to use the list instead of making a library (the list will
+            be more efficient and easier to use).
+
+        asn_exptypes : list of str, optional
```

Comment:
There are a few options:
- if the input data is an association, load a `ModelLibrary` for each `exptype` (creating the library is cheap as long as the members have a `group_id`)
- if the input data is an existing `ModelLibrary` (with multiple `exptype`s) the indexes corresponding to each `exptype` can be looked up by iterating through the members list via `library.asn['products'][0]['members']`
- if there's any chance the `exptype` of the models may have changed, the most sure-fire way would be to iterate through the library (or use `map_function`) to split the library

Only the first option above would produce a `ModelLibrary` for each `exptype` (the other options would results in lists of models). It might be possible to extend the `ModelLibrary` adding some sort of splitting API. However I'm not aware of any uses in `calwebb_image3` or `romancal` (the current targets). Hopefully we can extend the `ModelLibrary` uses to other portions of jwst (which will need this splitting) but I suggest we wait until we need the functionality to add it and perhaps we only add it in jwst (since it's not needed in romancal).